### PR TITLE
Automatic differentiation for lax.linalg.eig

### PR DESCRIPTION
**Pull request by jaj42**
The implementation was generated with Claude Code after providing the following resources:
- pytorch implementation of eig autodiff
- autograd implementation of eig autodiff
- fmmax implementation of eig autodiff
- Github issue #2748 discussion
- Giles, M. (2008). https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf
- Boeddeker, C. et al. (2017) https://arxiv.org/abs/1701.00392

I do not have the expertise to understand the underlying mathematics but I needed the implementation and thought it may be valuable for others.

I tested gradients against lax.linalg.eigh and jax.scipy.linalg.expm with identical results.

I also tested the implementation from within PyMC models with MCMC sampling and thousands of evaluations and obtained correct fast model convergence.

### CODE REVIEW

**Comment by mattjj** [comment]
@j-towns any chance you can take a look at this?

**Comment by mattjj** [comment]
@jaj42 can you add numerical tests, e.g. using `check_grads` or similar? (I mean, ask Claude to do it :P)

**Comment by mattjj** [comment]
I thought a bit about the phase ambiguity discussed in #2748. We want to rule out the case that the computed gradients depend on the arbitrary choice of phase in the eigendecomposition (only phase because the computed eigenvectors have unit norm). It'd be sufficient to rule out the downstream computation depending on the phases of the eigenvectors, but a weaker (ie better) condition is that the covector we try to pull back has the space of ambiguous tangent variations in its kernel. Instead of actually checking that dynamically (i.e. at runtime), we can just have the user promise it by passing an opt-in argument to `eig` (just like we do with `holomorphic=True` on `grad`).

For extra concreteness, consider the 1x1 case, i.e. the eigendecomposition of $x \in \mathbb{C}^{1 \times 1}$ into $a$ and $v$ such that $x = v a v^{-1}$. We could take $a = x$ and $v = 1$, but we could also take $v = e^{i \theta}$ for any $\theta \in \mathbb{R}$. That's the ambiguity! More generally, for $x \in \mathbb{C}^{n \times n}$, we have a vector of phases, one for each eigenvector.

When does that ambiguity not affect the results of an AD computation? Say we're using AD to compute something of the generic form $\langle u, \partial f(x) v \rangle$ where $\langle \cdot, \cdot \rangle$ is the standard inner product $\langle A, B \rangle = \text{Re} \text{Tr} ( A^H B )$ (because we're considering $\mathbb{R}$-linearity, i.e. $\mathbb{C}^{m \times n}$ as a real vector space). Moreover say $f = h \circ g$, so $\langle u, \partial f(x) v \rangle = \langle u \partial h(g(x)), \partial g(x) v \rangle$, where the output of $g$ has the ambiguity that we're worried about. We just need the ambiguous variation to be in the kernel of $\langle u \partial h(g(x)), \cdot \rangle$.

Because here the primal ambiguity is of the form  $v \mapsto v e^{i \theta}$ for any $\theta$, the tangent space of variations is $i \theta v$ (because $e^{i \theta} \approx 1 + i \theta$). So to rule out this ambiguity affecting the AD computation, we'd need $0 = \langle \bar V, i \theta V \rangle = \text{Re} i \text{Tr} ( \bar V^H V \text{diag}^* \theta )$ for all $\theta$ vectors, or essentially that $\text{Im} \text{diag} (V^H V) = 0$.

That's the condition that we could check at runtime, in the backward pass. But instead let's just provide an argument so the user can promise it's fine. (Maybe we can run the check if we detect concrete values are available, i.e. if we're in eager mode, or something.)

**Comment by j-towns** [comment]
@jaj42 to summarize what @mattjj is asking for:
 - Add a new kwarg (could be called `allow_eigvec_deriv`) to the `eig` primitive, default to `False`
 - Add explanation to the `eig` docstring that this kwarg is analogous to `holomorphic` on grad, and that it's there to assert that the user understands that the eigenvector derivative is only defined up to gauge, and that any gauge-dependent component is projected out. This is safe when what is downstream of eig, in the user's function, does not have any dependence on the phase/gauge.
 - In the `eig` `jvp`, when the new kwarg is `False` and derivatives w.r.t. the eigenvectors are required, raise a clear error
 - When the kwarg is True, return the JVP with the full gauge direction projected out (the PR already zeros the norm component via Re diag(Vᴴ dV); extending this to the full complex diagonal also zeros the phase component). This is a no-op if the user's function has no dependence on gauge, but makes the result well-defined regardless.

Is that a fair summary @mattjj? I'm also happy to do this (can push to the tip of the PR branch) if helpful.

**Comment by jaj42** [comment]
@mattjj @j-towns I hope the changes I pushed address your concerns. I reran the tests and they pass.

**Comment by mattjj** [comment]
@j-towns brilliantly explained, thanks for that!

File: jax/_src/lax/linalg.py [hunk 1]
```diff
@@ -169,15 +170,36 @@ def eig(
...
+  and 4.63 in https://arxiv.org/abs/1701.00392. Gradients through degenerate
+  (repeated) eigenvalues are mathematically ill-defined and will produce
+  numerically large or infinite values.
+
+  **Phase ambiguity in eigenvector gradients.** Eigenvectors are defined only
+  up to an arbitrary complex phase: if ``v`` is an eigenvector, so is
+  ``e^{iθ} v`` for any real ``θ``. This means eigenvector-dependent gradients
+  are only well-defined when the downstream computation does not depend on this
+  arbitrary phase choice. This is analogous to the ``holomorphic=True`` flag on
+  :func:`jax.grad`.
```
Comment (mattjj) [inline jax/_src/lax/linalg.py]:
```suggestion
  arbitrary phase choice.
```

File: jax/_src/lax/linalg.py [hunk 2]
```diff
@@ -169,15 +170,36 @@ def eig(
...
+
+  To prevent silently incorrect gradients, differentiation through eigenvectors
+  raises an error by default. Set ``allow_eigvec_deriv=True`` to opt in,
+  thereby asserting that your computation does not depend on the phase of the
```
Comment (mattjj) [inline jax/_src/lax/linalg.py]:
```suggestion
  thereby promising that your computation does not depend on the phase of the
```

File: jax/_src/lax/linalg.py [hunk 3]
```diff
@@ -169,15 +170,36 @@ def eig(
...
+  eigenvectors. When this flag is set, the JVP projects out the full gauge
+  ambiguity (both norm and phase components) from the eigenvector tangents,
+  making the result well-defined regardless.
```
Comment (mattjj) [inline jax/_src/lax/linalg.py:190]:
```suggestion
  ambiguity (both norm and phase components) from the eigenvector tangents.
```

File: jax/_src/lax/linalg.py [hunk 4]
```diff
@@ -169,15 +170,36 @@ def eig(
...
 
   Args:
     x: A batch of square matrices with shape ``[..., n, n]``.
     compute_left_eigenvectors: If true, the left eigenvectors will be computed.
     compute_right_eigenvectors: If true, the right eigenvectors will be
       computed.
+    allow_eigvec_deriv: If ``False`` (default), differentiating through
+      eigenvectors raises an error. Set to ``True`` to enable eigenvector
+      differentiation, asserting that your downstream computation does not
```
Comment (mattjj) [inline jax/_src/lax/linalg.py:199]:
```suggestion
      differentiation, promising that your downstream computation does not
```

File: jax/_src/lax/linalg.py (hunk 2 above)
Comment (mattjj) [reply jax/_src/lax/linalg.py]: We could add brief examples of what it looks like to depend vs not depend. Like, `eig(x)[1].sum()` is bad, while `s, v = eig(x); y = v @ jnp.diag(f(s)) @ jnp.linalg.inv(v)` is okay. (I'm not sure those are right; just a guess.)

Or we could leave it...

**Review by mattjj — COMMENTED** [review]
A couple docstring comments. I haven't checked the math yet, though I guess an LLM can do that at least as well as I can...

**Comment by mattjj** [comment]
An idea suggested by both Claude (in the tests) and @dougalm: why don't we just normalize the phase, like _normalizing_eig does in the tests? That is, take that as part of the definition of the primitive `eig`.

**Comment by mattjj** [comment]
https://github.com/jax-ml/jax/pull/36700#issuecomment-4246639433

I think we could "fix the gauge" like that, but it'd necessarily introduce singularities into our AD rules.

**Comment by j-towns** [comment]
The gauge is already fixed (see https://github.com/jax-ml/jax/issues/2748#issuecomment-627410072 and the comment below). I thought it would be awkward to support the fixed gauge directly but I pointed my LLM of choice at the problem and it seems it isn't too bad (https://github.com/jax-ml/jax/pull/36832, in particular the PR description). @mattjj maybe we should just go for the full support since it isn't too bad. WDYT?

File: tests/linalg_test.py [hunk 5]
```diff
@@ -420,6 +451,45 @@ def testEigvalsGrad(self, shape, dtype):
...
+    if not (compute_left_eigenvectors or compute_right_eigenvectors):
+      # Eigenvalue-only gradients are already covered by testEigvalsGrad.
+      self.skipTest("eigenvalue-only case covered by testEigvalsGrad")
+    rng = jtu.rand_default(self.rng())
+    a = rng(shape, dtype)
+    f = partial(_normalizing_eig,
+                compute_left_eigenvectors=compute_left_eigenvectors,
+                compute_right_eigenvectors=compute_right_eigenvectors)
+    tol = 1e-4 if dtype in (np.float64, np.complex128) else 5e-2
+    jtu.check_grads(f, (a,), order=1, modes=['fwd', 'rev'], rtol=tol, atol=tol)
```
Comment (j-towns) [inline tests/linalg_test.py]: It should be possible to test order=2 for high precision inputs (similar to what the Cholesky test above does).

**Comment by mattjj** [comment]
Thanks for pushing on this @jaj42 ! I’ll close this since we merged the other pr that this one seeded.
